Revive read/write engine lock to guard operations against resets

tlrx opened the pull request:
Draft pull request that re-introduces the engine read/write lock to guard against engine resets.

It differs from #124635 on the following:
- uses the `engineMutex` for creating/closing engines
- uses the reentrant r/w lock for retaining engine instances and for resetting the engine
- acquires the reentrant read lock during refreshes to prevent deadlocks during resets
- add tests to ensure no deadlock when re-acquiring read lock in refresh listeners

To do:
- add test for concurrent resets and engine acquisitions
- replace `getEngine` by `withEngine()` in most places ?

tlrx committed:
Revert "Revert "Use read/write engine lock to guard operations against resets (#124635)" (#125915)"

This reverts commit 7fadeeb7c9b168db3a2a99f4ca3e4c7db41e8044.

tlrx committed:
add EngineReadWriteLock

tlrx committed:
acquire read lock before refresh

tlrx committed:
Use lock for resets only

elasticsearchmachine commented:
Pinging @elastic/es-distributed-indexing (Team:Distributed Indexing)

tlrx committed:
Add test for deadlock

henningandersen commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -4363,54 +4500,59 @@ assert getActiveOperationsCount() == OPERATIONS_BLOCKED
-                    synchronized (engineMutex) {
-                        if (newEngineReference.get() == null) {
-                            throw new AlreadyClosedException("engine was closed");
+            engineResetLock.readLock().lock(); // prevent engine resets during rollback
```
We seem to already be protected against that here - like in close, perhaps either comment on that. OK to leave in place, seems good for lock ordering if we also take the `mutex` (and thus I do not have to check whether we do).

tlrx replied:
I removed the read lock in 4a8186247796830137ad6cfde5e8aeff0d7c882d.

henningandersen commented on server/src/main/java/org/elasticsearch/index/engine/EngineResetLock.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * See {@link ReentrantReadWriteLock#getQueuedWriterThreads()}
+     */
+    public Collection<Thread> getQueuedWriterThreads() {
```
I wonder about the package belonging of this class, seems more `IndexShard` related than `Engine`, since it protects something in `IndexShard` primarily. If we move this to `shard` then this can be package private too - I think it is only used in tests.

tlrx replied:
Makes sense, I pushed 4fa503c01d191cbdb41ad123e5f3e3973c4e48dc

henningandersen reviewed (commented):
This direction looks good. Only reviewed production parts for now. Left a few comments but nothing major.

tlrx committed:
sometimes

tlrx committed:
doc about non-blocking refreshes and reset

tlrx committed:
assertNoEngineResetLock

tlrx committed:
readLock and comment

tlrx committed:
move EngineResetLock

tlrx commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1807,38 +1813,43 @@ public CacheHelper getReaderCacheHelper() {
 
     public void close(String reason, boolean flushEngine, Executor closeExecutor, ActionListener<Void> closeListener) throws IOException {
         synchronized (engineMutex) {
+            engineResetLock.readLock().lock(); // prevent engine resets while closing
```
I kept the read lock and added a comment in 4a8186247796830137ad6cfde5e8aeff0d7c882d, let me know what you think.

tlrx commented on server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -2051,12 +2051,30 @@ protected final RefreshResult refresh(String source, SearcherScope scope, boolea
+                        }
                     } else {
-                        refreshed = referenceManager.maybeRefresh();
+                        if (engineReadLock.tryLock()) {
```
Makes sense, I pushed a bit of documentation in a80d07b8c785d706a8116f5bed84e39b6a29f26f.

tlrx commented on server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -2051,12 +2051,30 @@ protected final RefreshResult refresh(String source, SearcherScope scope, boolea
+
+                    // The shard uses a reentrant read/write lock to guard again engine changes, a type of lock that prioritizes the threads
+                    // waiting for the write lock over the threads trying to acquire a (non-reentrant) read lock. Because refresh listeners
+                    // are accessing the engine read lock, we need to ensure that they won't block if another thread is waiting for the
```
I pushed a6cad1945b98e5efdbc46903cee182f0238fe75d

tlrx commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1807,38 +1813,43 @@ public CacheHelper getReaderCacheHelper() {
 
     public void close(String reason, boolean flushEngine, Executor closeExecutor, ActionListener<Void> closeListener) throws IOException {
         synchronized (engineMutex) {
```
Done in 34c422b54f03eeec010ece29db99d3c685957c16

Commit:
[CI] Auto commit changes from spotless

tlrx committed:
== false

tlrx committed:
remove more readlocks

tlrx committed:
remove some assertNoEngineResetLock

tlrx committed:
also read lock on flushes

tlrx commented on server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -2191,83 +2209,94 @@ protected void flushHoldingLock(boolean force, boolean waitIfOngoing, ActionList
-        final long startTime = System.nanoTime();
+        // Acquire an engine read lock before the flush lock. If we were not acquiring a read lock here, a concurrent engine reset could
+        // hold the engine write lock and later be blocked waiting for the flush lock (still holding the write lock), while the current
+        // thread could be blocked waiting for the write lock to be released (and therefore never release the flush lock).
```
It's been detected by a test failure, see this thread dump:

```


Found one Java-level deadlock:
=============================
"elasticsearch[node_t3][clusterApplierService#updateTask][T#1]":
  waiting for ownable synchronizer 0x00000000e1ff06e0, (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync),
  which is held by "elasticsearch[node_t3][generic][T#2]"

"elasticsearch[node_t3][generic][T#2]":
  waiting for ownable synchronizer 0x00000000e3a10a30, (a java.util.concurrent.locks.ReentrantLock$NonfairSync),
  which is held by "elasticsearch[node_t3][force_merge][T#1]"

"elasticsearch[node_t3][force_merge][T#1]":
  waiting for ownable synchronizer 0x00000000e1ff06e0, (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync),
  which is held by "elasticsearch[node_t3][generic][T#2]"

Java stack information for the threads listed above:
===================================================
"elasticsearch[node_t3][clusterApplierService#updateTask][T#1]":
	at jdk.internal.misc.Unsafe.park(java.base@24/Native Method)
	- parking to wait for  <0x00000000e1ff06e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(java.base@24/LockSupport.java:223)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@24/AbstractQueuedSynchronizer.java:789)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireShared(java.base@24/AbstractQueuedSynchronizer.java:1118)
	at java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.lock(java.base@24/ReentrantReadWriteLock.java:739)
	at org.elasticsearch.index.shard.IndexShard.getEngine(IndexShard.java:3350)
	at org.elasticsearch.index.shard.IndexShard.commitStats(IndexShard.java:1374)
	at org.elasticsearch.index.IndexService.getNodeMappingStats(IndexService.java:356)
	at co.elastic.elasticsearch.stateless.autoscaling.memory.ShardsMappingSizeCollector.updateMappingMetricsForAllIndices(ShardsMappingSizeCollector.java:277)
	at co.elastic.elasticsearch.stateless.autoscaling.memory.ShardsMappingSizeCollector.clusterChanged(ShardsMappingSizeCollector.java:205)
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListener(ClusterApplierService.java:594)
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:579)
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:538)
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:460)
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:159)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:977)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:218)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:184)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@24/ThreadPoolExecutor.java:1095)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@24/ThreadPoolExecutor.java:619)
	at java.lang.Thread.runWith(java.base@24/Thread.java:1460)
	at java.lang.Thread.run(java.base@24/Thread.java:1447)
"elasticsearch[node_t3][generic][T#2]":
	at jdk.internal.misc.Unsafe.park(java.base@24/Native Method)
	- parking to wait for  <0x00000000e3a10a30> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(java.base@24/LockSupport.java:223)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@24/AbstractQueuedSynchronizer.java:789)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@24/AbstractQueuedSynchronizer.java:1029)
	at java.util.concurrent.locks.ReentrantLock$Sync.lock(java.base@24/ReentrantLock.java:154)
	at java.util.concurrent.locks.ReentrantLock.lock(java.base@24/ReentrantLock.java:323)
	at org.elasticsearch.index.engine.InternalEngine.flushHoldingLock(InternalEngine.java:2220)
	at co.elastic.elasticsearch.stateless.engine.IndexEngine.flushHoldingLock(IndexEngine.java:336)
	at org.elasticsearch.index.engine.Engine.flush(Engine.java:1334)
	at co.elastic.elasticsearch.stateless.engine.IndexEngine.flushHollow(IndexEngine.java:398)
	at co.elastic.elasticsearch.stateless.engine.IndexEngine.prepareForEngineReset(IndexEngine.java:458)
	at org.elasticsearch.index.shard.IndexShard.resetEngine(IndexShard.java:4452)
	- locked <0x00000000e3a10a20> (a java.lang.Object)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction.lambda$handleStartRelocationWithFreshClusterState$13(TransportStatelessPrimaryRelocationAction.java:305)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction$$Lambda/0x000076bfcbae6dd0.accept(Unknown Source)
	at org.elasticsearch.index.shard.IndexShard$2.onResponse(IndexShard.java:852)
	at org.elasticsearch.index.shard.IndexShard$2.onResponse(IndexShard.java:803)
	at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:413)
	at org.elasticsearch.index.shard.IndexShardOperationPermits$1.doRun(IndexShardOperationPermits.java:119)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:316)
	at org.elasticsearch.index.shard.IndexShardOperationPermits.waitUntilBlocked(IndexShardOperationPermits.java:94)
	at org.elasticsearch.index.shard.IndexShardOperationPermits.blockOperations(IndexShardOperationPermits.java:90)
	at org.elasticsearch.index.shard.IndexShard.relocated(IndexShard.java:803)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction.lambda$handleStartRelocationWithFreshClusterState$14(TransportStatelessPrimaryRelocationAction.java:282)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction$$Lambda/0x000076bfcbae5d78.accept(Unknown Source)
	at org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:261)
	at org.elasticsearch.action.support.SubscribableListener$SuccessResult.complete(SubscribableListener.java:394)
	at org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:314)
	at org.elasticsearch.action.support.SubscribableListener.addListener(SubscribableListener.java:210)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction.handleStartRelocationWithFreshClusterState(TransportStatelessPrimaryRelocationAction.java:276)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction$1.accept(TransportStatelessPrimaryRelocationAction.java:222)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction$1.accept(TransportStatelessPrimaryRelocationAction.java:219)
	at org.elasticsearch.indices.recovery.PeerRecoverySourceClusterStateDelay.ensureClusterStateVersion(PeerRecoverySourceClusterStateDelay.java:42)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction.handleStartRelocation(TransportStatelessPrimaryRelocationAction.java:210)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction.lambda$new$2(TransportStatelessPrimaryRelocationAction.java:166)
	at co.elastic.elasticsearch.stateless.recovery.TransportStatelessPrimaryRelocationAction$$Lambda/0x000076bfcb7872d8.messageReceived(Unknown Source)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:90)
	at org.elasticsearch.transport.InboundHandler.doHandleRequest(InboundHandler.java:289)
	at org.elasticsearch.transport.InboundHandler$1.doRun(InboundHandler.java:302)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1044)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@24/ThreadPoolExecutor.java:1095)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@24/ThreadPoolExecutor.java:619)
	at java.lang.Thread.runWith(java.base@24/Thread.java:1460)
	at java.lang.Thread.run(java.base@24/Thread.java:1447)
"elasticsearch[node_t3][force_merge][T#1]":
	at jdk.internal.misc.Unsafe.park(java.base@24/Native Method)
	- parking to wait for  <0x00000000e1ff06e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(java.base@24/LockSupport.java:223)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@24/AbstractQueuedSynchronizer.java:789)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireShared(java.base@24/AbstractQueuedSynchronizer.java:1118)
	at java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.lock(java.base@24/ReentrantReadWriteLock.java:739)
	at org.elasticsearch.index.engine.InternalEngine.refresh(InternalEngine.java:2063)
	at org.elasticsearch.index.engine.InternalEngine.flushHoldingLock(InternalEngine.java:2253)
	at co.elastic.elasticsearch.stateless.engine.IndexEngine.flushHoldingLock(IndexEngine.java:336)
	at org.elasticsearch.index.engine.Engine.flush(Engine.java:1334)
	at org.elasticsearch.index.engine.Engine.flush(Engine.java:1315)
	at org.elasticsearch.index.engine.InternalEngine.forceMerge(InternalEngine.java:2457)
	at co.elastic.elasticsearch.stateless.engine.IndexEngine.forceMerge(IndexEngine.java:693)
	at org.elasticsearch.index.shard.IndexShard.forceMerge(IndexShard.java:1575)
	at org.elasticsearch.action.admin.indices.forcemerge.TransportForceMergeAction.lambda$shardOperation$3(TransportForceMergeAction.java:106)
	at org.elasticsearch.action.admin.indices.forcemerge.TransportForceMergeAction$$Lambda/0x000076bfcbc0d938.get(Unknown Source)
	at org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:58)
	at org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:55)
	at org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:101)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1044)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@24/ThreadPoolExecutor.java:1095)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@24/ThreadPoolExecutor.java:619)
	at java.lang.Thread.runWith(java.base@24/Thread.java:1460)
	at java.lang.Thread.run(java.base@24/Thread.java:1447)
```

tlrx replied:
I should have precised that previously we were executing `prepareForEngineReset` under a read lock. We made it that way to avoid the issue with refresh listeners accessing the engine. 

We later noticed that it wasn't enough, since refresh listeners are also called when creating engines under the write lock, so we decided to acquire a read lock before the refresh lock.

With this last change, `prepareForEngineReset` can be executed under the write lock since it is garanteed that no other thread would block the refresh lock. But it now expose a similar issue with flushes, so I applied the same solution as for refresh: acquiring a read lock before flushes.

tlrx commented:
Thanks for your feedback @henningandersen. I addressed your comments, removed a bit more read lock acquisitions that were unnecessary thanks to the `engineMutex`, and also adds a read lock acquisition before flushes.

I'm investigating the serverless build failures, but otherwise CI is happy here.

henningandersen reviewed (approved):
LGTM.

tlrx committed:
feedback

tlrx commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -5094,6 +5099,344 @@ public void testRecordsForceMerges() throws IOException {
+            safeGet(close);
+
+            assertThat(primary.state(), equalTo(IndexShardState.CLOSED));
+            expectThrows(AlreadyClosedException.class, () -> primary.getEngine());
```
Makes sense, I pushed 94517c74fb20c35f0efdf765165732003aac7fca

Commit:
[CI] Auto commit changes from spotless

tlrx committed:
assert

tlrx commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2294,7 +2311,7 @@ public void performRecoveryRestart() throws IOException {
     public void resetRecoveryStage() {
         assert routingEntry().recoverySource().getType() == RecoverySource.Type.PEER : "not a peer recovery [" + routingEntry() + "]";
-        assert currentEngineReference.get() == null;
+        assert getEngineOrNull() == null;
```
Good point, I think we can for most places so I pushed 19683eb12a118e02fd4fa641aacde269dcfd7d98.

tlrx committed:
Merge branch 'main' into ES-10826-the-return

tlrx committed:
Merge branch 'main' into ES-10826-the-return

tlrx committed:
Merge branch 'main' into ES-10826-the-return

tlrx committed:
Merge branch 'main' into ES-10826-the-return

fcofdez commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -5094,6 +5099,348 @@ public void testRecordsForceMerges() throws IOException {
+        closeShards(shard);
+    }
+
+    public void testReentrantEngineReadLockAcquisitionInRefreshListener() throws Exception {
```
Great job on this test 👍

fcofdez reviewed (approved):
LGTM. Great job 🎉. I left a few minor comments, feel free to tackle them or not 👍.

tlrx committed:
onSettingsChanged

tlrx committed:
sameInstance

tlrx commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2623,9 +2640,14 @@ boolean shouldRollTranslogGeneration() {
-        Engine engineOrNull = getEngineOrNull();
-        if (engineOrNull != null) {
-            engineOrNull.onSettingsChanged();
+        engineResetLock.readLock().lock();
```
I pushed 2411a9b08bb48089dabe83b9320cc48d9629c9f7 to use withEngine here.

tlrx replied:
... and just reverted, since withEngine ensure we're not calling from the cluster state applier thread.

tlrx commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -5094,6 +5099,348 @@ public void testRecordsForceMerges() throws IOException {
+        holdEngineThread.start();
+        safeGet(hold);
+
+        assertThat(shard.getEngine(), instanceOf(InternalEngine.class));
```
Makes sense, I pushed 93c25daf97f4925bbe830b9247ca7909d6b0aedb

tlrx committed:
Merge branch 'main' into ES-10826-the-return

tlrx committed:
Revert "onSettingsChanged"

This reverts commit 2411a9b08bb48089dabe83b9320cc48d9629c9f7.

tlrx commented:
Thanks @henningandersen and @fcofdez !
